### DISCUSSION TITLE
Make Tool and Function classes easier to deserialize

### DISCUSSION THREAD

**Discussion by endeffects** [RageAgainstThePixel/OpenAI-DotNet#238] [Q&A]
Currently it is not possible to deserialize the Tool and Function classes by using System.Text.json.
My goal here is to load the tools / functions at runtime to enable / disable them on my need without creating a new release i.e.

I also have some other related points here:
1. It would be nice to have a more type safe implementation for the Function class that does not require json types.
2. It also would be nice not to have sealed classes here, to build additional forward logic on the openAi callback.
3. An alternative approach to keep the sealed classes would be the possiblity to declare a deserialisation type for the function arguments.

With that option it would be possible to dynamically deserialize the callback without any switches and pass it to some other processing handlers:

```csharp
        // Processing Function Callback from OpenAi
        
        if (source.FirstChoice.FinishReason != "tool_calls") return result;
        if (source.FirstChoice.Message?.ToolCalls == null) return result;
        
        foreach (var tool in source.FirstChoice.Message.ToolCalls)
        {
            var arguments = $"{tool.Function.Arguments}";
            
            // At the moment i have to use an additional Dictionary to get the correct Deserialization Type
            // In my case i'm deserializing the function arguments into types that are implementing IRequest<TResult> of the MediatR nuget package
            if (JsonSerializer.Deserialize(arguments, this.tools[tool.Function.Name].Type) is not IRequest<MessageModel> request) continue;

            // So the better approach would be something like:
            // if (JsonSerializer.Deserialize(arguments, tool.Function.ArgumentsType) is not IRequest<MessageModel> request) continue;

            // Then i can pass the request to the mediator that calls the required handler to process the function callback
            var result = await this.mediator.Send(request, cancellationToken);
        }
```

**Comment by StephenHodgson** [comment]
You're going to have to give a better explanation of your desired use case.

> Also please consider to use a more type safe implementation here and get rid of the json types.

The goal here is to let you unpack the objects as you need them. You should be able to deserialize any data you need already, since it is json, which means it is inherently serializable.

**Reply by ADefWebserver** [reply]
@endeffects  Can we get an example?

**Reply by endeffects** [reply]
i have updated my post and provided some more details and samples

**Comment by endeffects** [comment]
Here is a non working example on net core 8.0. 

```csharp
var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = false };        
using var stream = File.OpenRead("functions.json");        
var tools = JsonSerializer.DeserializeAsync<List<Tool>>(stream, options);
```

```json
[
  {
    "Function": {
      "Name": "FindSomething",
      "Description": "Find something",
      "Parameters": {
        "Type": "object",
        "Properties": {
          "Query": {
            "Type": "string",
            "Description": "The query"
          }
        },
        "Required": [ "query" ]
      }
    }
  }
]
```

**Reply by ADefWebserver** [reply]
@endeffects - Thank you that clarifies exactly what you are asking :)

This is @StephenHodgson decision, but my $0.02:

- Yeah you have to manually deserialize it. A pain but it will work.

_1. It would be nice to have a more type safe implementation for the Function class that does not require json types._
**[I don't think this is up to us. We can't create a 'contract' that OpenAI can break at any time]**

_2. It also would be nice not to have sealed classes here, to build additional forward logic on the openAi callback._
**[Again, this is to indicate something controlled by OpenAI]**

_3. An alternative approach to keep the sealed classes would be the possibility to declare a deserialization type for the function arguments._
**[But when OpenAI breaks it, it is still a breaking change, correct?]**

**Reply by endeffects** [reply]
Well, if openAi brakes something here everyone who uses callbacks has to fix it anyways, type safe or not. So I'm not sure if it is an argument.

**Reply by StephenHodgson** [reply]
@endeffects thanks for taking the time to clarify and update your post. I understand what you're trying to do now. 

The current implementation was a means to give developers the most flexibility by being able to deserialize their objects.

> type safe implementation

I don't believe you're using this term correctly, as type safety in .NET helps prevent objects of one type from accessing the memory assigned for another object. Which does not apply in this case, since we are using basic types to express the json values.

---

I've considered a few different ways to do to resolve this and make it a lot easier to define tools and functions in code.
Overall, I do want to overhaul this at some point to provide a better means to define and access Tools in a more structured and dynamic way. 

One idea I had was to add custom attributes to classes that define tools. or have an `ITool` interface you can simply add to class members. 

I also considered simply using a `dynamic` primitive type as well and having the runtime determine which to deserialize on its own.

I'm always open to suggestions here for implementation and it has been an area I've wanted to pick up and focus on at some point.

**Reply by StephenHodgson** [reply]
I also want to point out that technically the tool/functions _are_ able to be deserialized.
It just isn't as easy/straight forward as I'd like it to be.

```csharp
var functionArgs = JsonSerializer.Deserialize<WeatherArgs>(usedTool.Function.Arguments.ToString());
var functionResult = WeatherService.GetCurrentWeather(functionArgs);
```

**Reply by endeffects** [reply]
For the given example all properties are null on a deserialized function instance. So the issue is not just related to the json property types. 

I did not tested it for now, but maybe you need to set the JsonConstructor attribute on the parameterized constructor to fix this. 

Right now the property setters are inaccessible for the deserializer. 

https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/immutability?pivots=dotnet-5-0

**Reply by StephenHodgson** [reply]
> Right now the property setters are inaccessible for the deserializer.

This was intentional since objects returned by the remote resource should be immutable.

> For the given example all properties are null on a deserialized function instance.

If this was actually the case in practice then my unit tests would fail 😂

I believe that the function args themselves are not immutable

https://github.com/RageAgainstThePixel/OpenAI-DotNet/blob/main/OpenAI-DotNet-Tests/TestServices/WeatherArgs.cs

**Reply by StephenHodgson** [reply]
I think for now as a stop gap an extension method to deserialize, might prove useful.

```csharp
var functionArgs = usedTool.GetArgs<T>();
var functionResult = FunctionClass.FunctionCall(functionArgs);
```

**Reply by endeffects** [reply]
Which tests are you referencing to? I did not found one that uses json strings. The one I saw were using dynamic types. 

Immutable does not mean undeserializable? It means that the instance cannot be modified, but it can be cloned for sure.

**Reply by StephenHodgson** [reply]
You got me confused. I've explicitly kept the tool/function calls as json. Just bc it's in the json object type doesn't make it any less serializable. You still need to deserialize any of these objects to a specific type in order to use them.

I suppose you are concerned about this member having the `dynamic` signature?
https://github.com/RageAgainstThePixel/OpenAI-DotNet/blob/8e99b079381f87c5d056a503ae852065f6e09714/OpenAI-DotNet/Chat/Message.cs#L89

This is mainly due to the fact that the API returns different objects depending on if sending or receiving and I didn't feel like adding fancy support for it or forcing people to upgrade to .net 8+

**Reply by endeffects** [reply]
I just said that i still cannot deserialize the tool nor the function class from a pure json string as shown in my previous example. You said that this cannot be true because all your tests would fail in this case. So I wonder which tests are you referring too.

**Reply by endeffects** [reply]
Just to be clear here, we are discussing two different things in parallel right now. 

1. i want to be able to deserialize the tool and function classes 

And unrelated to this one:

2. I want to define a deserializion type somewhere for the function arguments

**Reply by StephenHodgson** [reply]
All my unit tests are based off of deserializing the JSON string 😂

**Reply by StephenHodgson** [reply]
1. Is already possible
2. Is something I've been wanting to tackle but haven't had the time to get around to it

**Reply by StephenHodgson** [reply]
I guess I'll need a more concrete example of what exactly your use case is, because I still don't quite understand what the problem is for 1.

**Reply by endeffects** [reply]
i've checked that deserialization issue again and it was my fault. you are using JsonPropertyName attributes that are enforcing a case sensitivity over the PropertyNameCaseInsensitive setting on the serializer options. in my sample json the property names were title case, so the values got ignored.

**Comment by ADefWebserver** [comment]
> Well, if openAi brakes something here everyone who uses callbacks has to fix it anyways, type safe or not. So I'm not sure if it is an argument.

Ok I see your point. So is the request now this?:

`3. An alternative approach to keep the sealed classes would be the possibility to declare a deserialization type for the function arguments.`

**Reply by endeffects** [reply]
Yes, this would allow us to use an implicit deserialisation instead of an explicit one for the argument types.

**Reply by StephenHodgson** [reply]
I considered adding an abstract base tool class or interface for exactly this reason.

Unsure which would be the better solution in this case but the polymorphic nature of some of the function calls made that a bit harder to do in practice.

**Reply by endeffects** [reply]
Maybe an alternative would be a generic function class that inherits from the existing one and can be passed to the tool class. 

In that way all your existing code would stay as it is and I can keep my inheriting collection of tools. 

`public class Function<TArgs> : Function
{
    public TArgs Type { get; }
}
`

Maybe an Interface to upcast and access the type property is required too.

**Reply by StephenHodgson** [reply]
Draft PR https://github.com/RageAgainstThePixel/OpenAI-DotNet/pull/241
